openedx/core/djangoapps/content_libraries/libraries_index.py

```diff
@@ -0,0 +1,159 @@
+    """
+    pass
+
+
+class ContentLibraryIndexer:
+    """
+    Class to perform indexing for blockstore-based content libraries
+    """
+
+    INDEX_NAME = "content_library_index"
+    LIBRARY_DOCUMENT_TYPE = "content_library"
+
+    @classmethod
+    def index_libraries(cls, library_keys):
+        """
+        Index the specified libraries. If they already exist, replace them with new ones.
+        """
+        from openedx.core.djangoapps.content_libraries.api import DRAFT_NAME
+
+        searcher = SearchEngine.get_search_engine(cls.INDEX_NAME)
+
+        library_dicts = []
+
+        for library_key in library_keys:
+            ref = ContentLibrary.objects.get_by_key(library_key)
+            lib_bundle = LibraryBundle(library_key, ref.bundle_uuid, draft_name=DRAFT_NAME)
+            num_blocks = len(lib_bundle.get_top_level_usages())
+            last_published = lib_bundle.get_last_published_time()
+            last_published_str = None
+            if last_published:
+                last_published_str = last_published.strftime('%Y-%m-%dT%H:%M:%SZ')
+            (has_unpublished_changes, has_unpublished_deletes) = lib_bundle.has_changes()
+
+            bundle_metadata = get_bundle(ref.bundle_uuid)
+
+            library_dict = {
+                "id": str(library_key),
+                "uuid": str(bundle_metadata.uuid),
+                "title": bundle_metadata.title,
+                "description": bundle_metadata.description,
+                "num_blocks": num_blocks,
+                "version": bundle_metadata.latest_version,
+                "last_published": last_published_str,
+                "has_unpublished_changes": has_unpublished_changes,
+                "has_unpublished_deletes": has_unpublished_deletes,
+            }
+            library_dicts.append(library_dict)
+
+        return searcher.index(cls.LIBRARY_DOCUMENT_TYPE, library_dicts)
+
+    @classmethod
+    def get_libraries(cls, library_keys):
+        """
+        Retrieve a list of libraries from the index
+        """
+        searcher = SearchEngine.get_search_engine(cls.INDEX_NAME)
+        library_keys_str = [str(key) for key in library_keys]
+        response = searcher.search(
+            doc_type=cls.LIBRARY_DOCUMENT_TYPE,
+            field_dictionary={"id": library_keys_str},
+            size=MAX_SIZE,
+        )
+
+        # Search results may not retain the original order of keys - we use this
+        # dict to construct a list in the original order of library_keys
+        response_dict = {
+            result["data"]["id"]: result["data"]
+            for result in response["results"]
+        }
+        if len(response_dict) != len(library_keys_str):
+            missing = set(library_keys_str) - set(response_dict.keys())
+            raise LibraryNotIndexedException("Keys not found in index: {}".format(missing))
+        return [
+            response_dict[key]
+            for key in library_keys_str
+        ]
```

Comment:
Logic looks sound, but it also makes the cache retrieval all-or-nothing. 

Would it ever be the case that *some* libraries are missing from the index? If so, would it be worth returning the libraries that we successfully in the index instead of raising? Callers such as `get_metadata_from_index` would have to then fill in the missing entries with that limited metadata.

Reply:
Similar to above, this behavior was changed  in #24510 [here](https://github.com/edx/edx-platform/pull/24510/commits/67f67ff7a18f49a2ae7e118516f6a1104bc5119c#diff-8cfdde6bceaaa7f5b1e1fbc912439f52L252) - where if a library doesn't exist in elastic - it is not returned (as we weren't able to differentiate between unindexed libraries vs them not existing in the filter supplied). lmk if this specific change needs to be backported to this commit.

Reply:
Great; no backport necessary here.